futures-util/src/stream/async_reader.rs

```diff
@@ -0,0 +1,157 @@
+use core::marker::Unpin;
+use core::pin::Pin;
+use futures_core::stream::Stream;
+use futures_core::task::{LocalWaker, Poll};
+use futures_io::AsyncRead;
+use std::cmp;
+use std::io::Result;
+
+/// A stream combinator which interprets a stream of bytes as an `AsyncRead`.
+///
+/// This structure is produced by the `Stream::async_reader` method.
+#[derive(Debug)]
+#[must_use = "streams do nothing unless polled"]
+pub struct AsyncReader<St>
+where
+    St: Stream + Unpin,
+    St::Item: AsRef<[u8]>,
+{
+    stream: St,
+    state: ReadState<St::Item>,
+}
+
+impl<St: Stream + Unpin> Unpin for AsyncReader<St>
+where
+    St: Stream + Unpin,
+    St::Item: AsRef<[u8]>,
+{
+}
+
+#[derive(Debug)]
+enum ReadState<T: AsRef<[u8]>> {
+    Ready { chunk: T, chunk_start: usize },
+    NotReady,
+    Eof,
+}
+
+impl<St> AsyncReader<St>
+where
+    St: Stream + Unpin,
+    St::Item: AsRef<[u8]>,
+{
+    pub(super) fn new(stream: St) -> Self {
+        AsyncReader {
+            stream: stream,
+            state: ReadState::NotReady,
+        }
+    }
+}
+
+impl<St> AsyncRead for AsyncReader<St>
+where
+    St: Stream + Unpin,
+    St::Item: AsRef<[u8]>,
+{
+    fn poll_read(
+        &mut self,
+        lw: &LocalWaker,
+        buf: &mut [u8],
+    ) -> Poll<Result<usize>> {
+        loop {
+            match self.state {
+                ReadState::Ready {
+                    ref mut chunk,
+                    ref mut chunk_start,
+                } => {
+                    let chunk = chunk.as_ref();
+                    let len = cmp::min(buf.len(), chunk.len() - *chunk_start);
+
+                    buf[..len].copy_from_slice(
+                        &chunk[*chunk_start..*chunk_start + len],
+                    );
+                    *chunk_start += len;
+
+                    if chunk.len() == *chunk_start {
```

Comment:
In this case I think we need to go ahead and pull the next item from the stream and start loading it into the buffer as well.

Reply:
(but probably only do this if the buffer provided is big enough to hold more data)

Reply:
This is contrary to what things like `std::io::BufReader` do. If `poll_next` returns an error, what would be returned? The number of bytes that *were* read, or the error?

Reply:
Ack, speaking of errors, I really ought to be consuming a stream of `Result<AsRef<T>, std::io::Error>`s. I'll port my stuff over to it.

I also agree with @seanmonstar here. The only way I can think of that we could handle this case is if `AsyncRead` error includes the partial length that we read. I also don't know if there'd a real performance win if we were looping inside `AsyncReader` or an outer loop.

Reply:
I'd return the number of bytes that were read and store the error for the next time, but it also seems fine to me to not do this-- if the stdlib types don't typically do this, then I suppose it'd be more consistent not to.